File: java/src/main/java/edu/umn/nlpnewt/internal/Structs.java
/*
 * Copyright 2019 Regents of the University of Minnesota
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package edu.umn.nlpnewt.internal;

import com.google.protobuf.ListValue;
import com.google.protobuf.NullValue;
import com.google.protobuf.Struct;
import com.google.protobuf.Value;
import edu.umn.nlpnewt.Internal;
import edu.umn.nlpnewt.JsonObject;

import java.util.*;

@Internal
final class Structs {

  public static void copyStructToJsonObjectBuilder(Struct struct,
                                                   JsonObject.AbstractBuilder<?, ?> builder) {
    Map<String, Value> fieldsMap = struct.getFieldsMap();
    for (Map.Entry<String, Value> entry : fieldsMap.entrySet()) {
      builder.setProperty(entry.getKey(), getValue(entry.getValue()));
    }
  }

  public static void copyJsonObjectToStruct(JsonObject jsonObject, Struct.Builder structBuilder) {
    internalCopyJsonObject(jsonObject, structBuilder, new ArrayDeque<>());
  }

  private static void internalCopyJsonObject(JsonObject jsonObject,
                                             Struct.Builder struct,
                                             Deque<Object> parents) {
    parents.push(jsonObject);
    for (Map.Entry<?, ?> entry : jsonObject.entrySet()) {
      Object key = entry.getKey();
      if (!(key instanceof String)) {
        throw new IllegalArgumentException("Key is not a string key");
      }
      Object value = entry.getValue();

      struct.putFields((String) key, createValue(value, parents));
    }
    parents.pop();
  }

  private static Value createValue(Object from, Deque<Object> parents) {
    if (parents.contains(from)) {
      throw new IllegalArgumentException("Circular reference attempting to convert map to struct.");
    }
    if (from instanceof List) {
      parents.push(from);
      ListValue.Builder builder = ListValue.newBuilder();
      List fromList = (List) from;
      for (Object item : fromList) {
        Value value = createValue(item, parents);
        builder.addValues(value);
      }
      parents.pop();
      return Value.newBuilder().setListValue(builder).build();
    } else if (from instanceof JsonObject) {
      parents.push(from);
      Struct.Builder builder = Struct.newBuilder();
      JsonObject jsonObject = (JsonObject) from;
      internalCopyJsonObject(jsonObject, builder, parents);
      parents.pop();
      return Value.newBuilder().setStructValue(builder).build();
    } else if (from instanceof Double) {
      return Value.newBuilder().setNumberValue((Double) from).build();
    } else if (from instanceof String) {
      return Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build();
    } else {
      throw new IllegalArgumentException("Incompatible value type: " + from.getClass().getCanonicalName());
    }
  }

  static Object getValue(Value from) {
    switch (from.getKindCase()) {
      case NULL_VALUE:
        return null;
      case NUMBER_VALUE:
        return from.getNumberValue();
      case STRING_VALUE:
        return from.getStringValue();
      case BOOL_VALUE:
        return from.getBoolValue();
      case STRUCT_VALUE:
        JsonObject.Builder builder = new JsonObject.Builder();
        copyStructToJsonObjectBuilder(from.getStructValue(), builder);
        return builder.build();
      case LIST_VALUE:
        List<Object> list = new ArrayList<>();
        for (Value value : from.getListValue().getValuesList()) {
          list.add(getValue(value));
        }
        return list;
      case KIND_NOT_SET:
      default:
        throw new IllegalStateException("Unrecognized kind of struct value.");
    }
  }
}

File: python/mtap/_structs.py
# Copyright 2019 Regents of the University of Minnesota.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Internal utilities for interacting with protobuf structs."""
from typing import Sequence, Mapping

from google.protobuf import struct_pb2


def copy_dict_to_struct(d, struct, parents=None):
    if parents is None:
        parents = [d]
    for k, v in d.items():
        copy_item_to_struct(k, v, struct, parents)


def copy_item_to_struct(k, v, struct, parents):
    if isinstance(v, (str, float, bool, int)) or v is None:
        struct[k] = v
    elif isinstance(v, Sequence):
        _check_for_reference_cycle(v, parents)
        copy_list_to_list_value(v, struct.get_or_create_list(k), parents + [v])
    elif isinstance(v, Mapping):
        _check_for_reference_cycle(v, parents)
        copy_dict_to_struct(v, struct.get_or_create_struct(k), parents + [v])
    else:
        raise TypeError("Unrecognized type:", type(v))


def copy_list_to_list_value(lst, list_value, parents):
    for v in lst:
        if isinstance(v, (str, float, bool, int)) or v is None:
            list_value.append(v)
        elif isinstance(v, Sequence):
            _check_for_reference_cycle(v, parents)
            copy_list_to_list_value(v, list_value.add_list(), parents + [v])
        elif isinstance(v, Mapping):
            _check_for_reference_cycle(v, parents)
            copy_dict_to_struct(v, list_value.add_struct(), parents + [v])
        else:
            raise TypeError("Unrecognized type:", type(v))


def _check_for_reference_cycle(o, parents):
    for parent in parents:
        if o is parent:
            raise ValueError("Reference cycle while preparing for serialization.")


def copy_list_value_to_list(list_value, lst):
    for v in list_value:
        if isinstance(v, (str, float, bool, int)) or v is None:
            lst.append(v)
        elif isinstance(v, struct_pb2.ListValue):
            lst2 = []
            copy_list_value_to_list(v, lst2)
            lst.append(lst2)
        elif isinstance(v, struct_pb2.Struct):
            d = {}
            copy_struct_to_dict(v, d)
            lst.append(d)
        else:
            raise TypeError("Unrecognized type:", type(v))


def copy_struct_to_dict(struct, dictionary):
    for k, v in struct.items():
        if isinstance(v, (str, float, bool, int)) or v is None:
            dictionary[k] = v
        elif isinstance(v, struct_pb2.ListValue):
            lst = []
            copy_list_value_to_list(v, lst)
            dictionary[k] = lst
        elif isinstance(v, struct_pb2.Struct):
            d = {}
            copy_struct_to_dict(v, d)
            dictionary[k] = d
        else:
            raise TypeError("Unrecognized type:", type(v))


